Add Converters for Additional Types

SethFalco opened the pull request:
This adds the following converters, with test cases:
* Color 
* Dimension
* InetAddress
* Locale
* Pattern
* Point

You can see the test cases for example inputs/outputs.

I've just taken my PR from DeltaSpike (https://github.com/apache/deltaspike/pull/109) and converted the converters that didn't exist here to work with this project instead. This is just me acting on a comment in the PR (https://github.com/apache/deltaspike/pull/109#issuecomment-720469005) in case it does happen, I figured I could have the PR here already.

This does not port all changes from my other PR, for example I implemented some existing converters differently, but I'll PR those separately, so they can just be closed if they're unfavorable.

melloware reviewed (approved):
Very nice additions!

melloware commented:
Definitely submit your PR of your other changes.  I would like to see it and I did some of these converters and wasn't quite sure the best way to handle so i will be interested to see what you have done.

melloware commented:
Also rebase this one please.

garydgregory commented on src/main/java/org/apache/commons/beanutils2/converters/ColorConverter.java:
```diff
@@ -0,0 +1,216 @@
+    }
+
+    /**
+     * @param defaultValue The default value to be returned
```
Missing sentence in Javadoc: "Constructs..."

garydgregory commented on src/main/java/org/apache/commons/beanutils2/converters/ColorConverter.java:
```diff
@@ -0,0 +1,216 @@
+
+    /**
+     * <p>
+     *     Convert the configuration value to a Java {@link Color} object,
```
In general, I use "Convert..." -> "Converts..."

SethFalco commented:
Thanks for the comments, I believe both should be resolved now.
Let me know if there're any other issues.

garydgregory commented on src/main/java/org/apache/commons/beanutils2/converters/ColorConverter.java:
```diff
@@ -0,0 +1,216 @@
+    protected <T> T convertToType(Class<T> type, Object value) throws Throwable {
+        if (Color.class.isAssignableFrom(type)) {
+            Objects.requireNonNull(value, "Value can't be null.");
+            final String stringValue = value.toString();
```
Please rebase on master, then reuse the new convenience method in `AbstractConverter` to replace:
```
            Objects.requireNonNull(value, "Value must not be null.");
            final String stringValue = value.toString();
```
with:
```
final String stringValue = toString(value);
```

garydgregory commented on src/main/java/org/apache/commons/beanutils2/converters/DimensionConverter.java:
```diff
@@ -0,0 +1,112 @@
+    protected <T> T convertToType(Class<T> type, Object value) throws Throwable {
+        if (Dimension.class.isAssignableFrom(type)) {
+            Objects.requireNonNull(value, "Dimensions can not be null.");
+            final String stringValue = value.toString();
```
Please rebase on master, then reuse the new convenience method in `AbstractConverter` to replace:
```
            Objects.requireNonNull(value, "Value must not be null.");
            final String stringValue = value.toString();
```
with:
```
final String stringValue = toString(value);
```

garydgregory commented on src/main/java/org/apache/commons/beanutils2/converters/InetAddressConverter.java:
```diff
@@ -0,0 +1,85 @@
+    protected <T> T convertToType(Class<T> type, Object value) throws Throwable {
+        if (InetAddress.class.isAssignableFrom(type)) {
+            Objects.requireNonNull(value, "Value can't be null.");
+            final String stringValue = value.toString();
```
See above.

garydgregory commented on src/main/java/org/apache/commons/beanutils2/converters/LocaleConverter.java:
```diff
@@ -0,0 +1,78 @@
+        if (Locale.class.isAssignableFrom(type)) {
+            Objects.requireNonNull(value, "Value can't be null.");
+            final String stringValue = value.toString();
+
```
See above.

garydgregory commented on src/main/java/org/apache/commons/beanutils2/converters/PatternConverter.java:
```diff
@@ -0,0 +1,79 @@
+    protected <T> T convertToType(Class<T> type, Object value) throws Throwable {
+        if (Pattern.class.isAssignableFrom(type)) {
+            Objects.requireNonNull(value, "Regular expression can't be null.");
+            final String stringValue = value.toString();
```
See above.

garydgregory commented on src/main/java/org/apache/commons/beanutils2/converters/PointConverter.java:
```diff
@@ -0,0 +1,102 @@
+    protected <T> T convertToType(Class<T> type, Object value) throws Throwable {
+        if (Point.class.isAssignableFrom(type)) {
+            Objects.requireNonNull(value, "Value must not be null.");
+            final String stringValue = value.toString();
```
See above.

garydgregory reviewed (changes requested):
Hi @SethFalco 

Thank you for your PR.

Please rebase on master and see my comments scattered throughout which apply to the whole PR.

G

garydgregory commented on src/main/java/org/apache/commons/beanutils2/converters/ColorConverter.java:
```diff
@@ -0,0 +1,225 @@
+import java.util.Objects;
+
+/**
+ * <p>Converts a configuration property into a Java {@link Color} object.</p>
```
The first sentence in a Javadoc must end in a period but it does NOT need to be enclosed in a paragraph. Javadoc picks out the 1st sentence for the summary table automatically based on the period.

garydgregory commented on src/main/java/org/apache/commons/beanutils2/converters/PatternConverter.java:
```diff
@@ -0,0 +1,79 @@
+     * Constructs a {@link org.apache.commons.beanutils2.Converter} that will return
+     * the specified default value if a conversion error occurs.
+     *
+     * @param defaultValue The default value to be returned
```
For all files, the checktyle settings in this components allows for line length to 120, so use them :-)

garydgregory commented on src/main/java/org/apache/commons/beanutils2/ConvertUtilsBean.java:
```diff
@@ -516,10 +534,13 @@ private void registerStandard(final boolean throwException, final boolean defaul
         // @formatter:off
         register(Class.class,          throwException ? new ClassConverter()          : new ClassConverter(null));
+        register(ColorConverter.class, throwException ? new ColorConverter()          : new ColorConverter(null));
         register(Enum.class,           throwException ? new EnumConverter()           : new EnumConverter(null));
```
Isn't the first argument supposed to be Color? If so, how are the tests passing?

SethFalco commented on src/main/java/org/apache/commons/beanutils2/converters/ColorConverter.java:
```diff
@@ -0,0 +1,216 @@
+            Objects.requireNonNull(value, "Value can't be null.");
+            final String stringValue = value.toString();
+
+            switch (stringValue.toLowerCase()) {
```
Please let me know if there are any problems with the approach I'll be taking here.

I agree handling `Color#toString()` would be nice. While I was at it, I opted to support some terser versions since I was using regular expression anyway, and it wasn't much additional work:

Regex:
`^(?:[A-Za-z\d._]+)??\[?(?:r=)?(\d{1,3}),(?:g=)?(\d{1,3}),(?:b=)?(\d{1,3})\]?$`

`java.awt.Color[r=255,g=255,b=255]` - The normal `Color#toString()` result.  
`org.apache.ExtendedColor[r=255,g=0,b=255]` - If the `Color` class is extended, but `#toString()` is not overridden.  
`[r=0,g=255,b=0]` - Without the package.  
`r=64,g=64,b=64` - Without the backets.  
`255,175,175` - Without the color labels.

SethFalco commented:
Sorry for the delay in coming back to this.
Ended up getting pretty busy, then put this off for a while.

* I've rebased with master.
* Squashed my commits.
* I believe I've resolved all of your feedback.

High hopes everything is cool, but I'll be available if you have any further feedback.

SethFalco commented:
@garydgregory If this PR it too large to review at once, would you prefer if I split it up into smaller ones?

I could do one PR per converter for example?

garydgregory commented:
I'm sure it's fine for now (13 files).

Gary

On Mon, Apr 25, 2022, 07:28 Seth Falco ***@***.***> wrote:

> @garydgregory <https://github.com/garydgregory> If this PR it too large
> to review at once, would you prefer if I split it up into smaller ones?
>
> I could do one PR per converter for example?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/commons-beanutils/pull/47#issuecomment-1108449860>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJB6N6CVEQW2JMEWHTDT6LVGZ6VDANCNFSM4TIZJIZA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

garydgregory commented:
This is OK BUT I think we need to spit up this project into a multi-module Maven project because we are going to want to split out of module for code that depends on java.desktop at least. I know that's been a complaint in other Commons projects.

SethFalco commented:
Thanks for taking a look! And makes sense, in DeltaSpike there were also comments regarding the awt packages. 

Feel free to let me know how you want to proceed, and I don't mind dropping some of them from this PR.

(The last push just amends some of the commends/JavaDocs btw.)

garydgregory commented:
I don't think you need to do anything. I'll take a look later at how to best split things up.

garydgregory commented:
@SethFalco 
Run 'mvn' by itself to catch all these build errors.
